apps/pay/components/Memo/index.tsx

```diff
@@ -1,46 +1,45 @@
-import { useRouter } from "next/router"
 import React from "react"
 import { Modal } from "react-bootstrap"
 
-import { ACTIONS, ACTION_TYPE } from "../../pages/_reducer"
+import { useSearchParams, useParams } from "next/navigation"
+
+import { ACTIONS, ACTION_TYPE } from "../../app/reducer"
 
 import styles from "./memo.module.css"
 
+import { extractSearchParams } from "@/utils/utils"
+
 interface Props {
   state: React.ComponentState
   dispatch: React.Dispatch<ACTION_TYPE>
 }
 
 const Memo = ({ state, dispatch }: Props) => {
-  const router = useRouter()
-  const { username, amount, sats, unit, memo, display } = router.query
+  const searchParams = useSearchParams()
+  const { username } = useParams()
+  const { amount, sats, unit, memo, display } = extractSearchParams(searchParams)
+
   const [openModal, setOpenModal] = React.useState<boolean>(false)
   const [note, setNote] = React.useState<string>(memo?.toString() || "")
 
   const handleSetMemo = () => {
     if (unit === "SAT" || unit === "CENT") {
-      router.push(
-        {
-          pathname: `${username}`,
-          query: {
-            amount: amount,
-            sats: sats,
-            unit: unit,
-            memo: note,
-            display,
-          },
-        },
-        undefined,
-        { shallow: true },
+      window.history.pushState(
+        {},
+        "",
+        `${username}?${new URLSearchParams({
+          amount: amount,
+          sats: sats,
+          unit: unit,
+          memo: note,
+          display,
+        }).toString()}`,
       )
     } else {
-      router.push(
-        {
-          pathname: `${username}`,
-          query: { memo: note, display },
-        },
-        undefined,
-        { shallow: true },
+      window.history.pushState(
```

Comment:
ok I see this is a new feature of nextjs 14.1: https://nextjs.org/blog/next-14-1#windowhistorypushstate-and-windowhistoryreplacestate